Install approach FAQ (#8455)
istio/istio.io: content/en/faq/setup/install-method-selection.md

---
title: Which Istio installation method should I use?
weight: 10
---

In addition to the simple [getting started](/docs/setup/getting-started) evaluation install, there are several different
methods you can use to install Istio. Which one you should use depends on your production requirements.
The following lists some of the pros and cons of each of the available methods:

1. [istioctl install](/docs/setup/install/istioctl/)

>     The simplest and most qualified installation/management path with high security.

Reviewer: replace `/` with `and`

>     This is the community recommended method for most use cases.

Reviewer: the word community adds no value to the sentence. can you remove it?

    Pros:

    - Thorough configuration validation and health verification.
    - Uses the `IstioOperator` API which provides extensive configuration/customization options.
    - No in-cluster privileged pods needed. Changes are actuated by running the `istioctl` command.

    Cons:

>     - Multiple binaries must be managed, one per Istio minor version.

Reviewer: This isn't accurate, unless using the built-in manifests which is on a plan to deprecate. There is a flag to provide a URL to bring your own charts. @howardjohn you were working on the deprecation here, any word on progress?

>     - Some configurations may not be fully declarative and safe to use in CI pipelines (e.g., enabling third-party JWT).

Reviewer 2: I worry that this is vague enough to put people off and cause confusion. @nrjpoddar could you reiterate what you consider not declarative? The apply code path is actually identical to the operator, just the actuation is through running `istioctl install`.
Maybe a different way to put the high level concern is that the output manifest is directly applied as part of the command e.g. "If you need to generate and apply the output manifest through some other method (e.g. kubectl), consider using manifest generate or helm."

>       thereby producing varying installations in different Kubernetes environments.

Reviewer: instead of `varying installations` - how about `producing environmentally optimized installations`.

This is shorter and cleaner.

1. [Istio Operator](/docs/setup/install/operator/)

    Simple installation path without `istioctl` binaries. This is also a highly recommended approach unless

>     security issues cannot be addressed in the deployment environment.

Reviewer 3: ```suggestion
    for simplified upgrade workflows where running in-cluster privileged controller is not a concern.
```

    for simplified upgrade workflows where running an in-cluster privileged controller is not a concern.

    Pros:

>     - Same API as `istioctl install` but actuation is through a controller pod in the cluster with a fully declarative operation.

Reviewer: remove but and replace with period.

    - No need to manage multiple `istioctl` binaries.

    Cons:

>     - High privilege pod running in the cluster has security implications.

Reviewer 3: ```suggestion
    - High privilege controller running in the cluster has security risks.
```

>     - No need to manage multiple `istioctl` binaries.

Reviewer: Perhaps I am living in the future. I don't see the need to manage multiple `istioctl` binaries today. @ostromart please correct me if I am wrong?

>     - High privilege controller running in the cluster has security risks.

Reviewer: has -> poses

> 1. [istioctl manifest generate](/docs/setup/install/istioctl/#generate-a-manifest-before-installation)

Reviewer 2: alpha?

> 1. [istioctl manifest generate (alpha)](/docs/setup/install/istioctl/#generate-a-manifest-before-installation)

Reviewer: please remove alpha. This has been stable since 1.4 or 1.5. There has never been a report I am aware of that produces incorrect results.

    Generate the Kubernetes manifest and then apply with `kubectl apply --prune`.

>     This method is suitable where strict auditing or control of output manifests is needed.

Reviewer 3: ```suggestion
    This method is suitable where strict auditing or augmentation of output manifests is needed.
```

    This method is suitable where strict auditing or augmentation of output manifests is needed.

    Pros:

>     - Charts are generated from the same `IstioOperator` API as used in `istioctl install` and Operator.

Reviewer: 49 and 50 are shades of one another. Would you merge them?

    - Uses the `IstioOperator` API which provides extensive configuration/customization options.

> 

Reviewer: While I hate to dig into implementation details in a FAQ, I think it is important to note, somewhere in here, that all the `istioctl` based models use the same exact code for applying and managing the resulting K8s implementation.

    - Some checks performed in `istioctl install` and Operator are not done.
    - UX less streamlined compared to `istioctl install`.
    - Error reporting not as robust as `istioctl install` for apply step.

> 1. [install using Helm](/docs/setup/install/helm/)

Reviewer 2: alpha?

>     Using Helm charts allows easy integration with Helm based installation and automation pipelines.

Reviewer 3: ```suggestion
    Using Helm charts allows easy integration with Helm based workflows and automated resource pruning during upgrades. 
```

> 1. [install using Helm (alpha)](/docs/setup/install/helm/)

Reviewer 3: ```suggestion
1. [Install using Helm (alpha)](/docs/setup/install/helm/)
```

> 1. [install using Helm (alpha)](/docs/setup/install/helm/)

Reviewer: is `Helm` alpha or experimental for this release? I don't recall seeing an alpha checklist in the github.com/istio/enhancements repository. If its to be alpha, can you please execute the alpha checklist? Alpha requires (IIRC) TOC signoff, as well as any related WG leads to sign off.

Reviewer 3: It's alpha and has been discussed and reviewed by @ostromart. I will create an issue in enhancement repo for broader visibility.

Reviewer 3: https://github.com/istio/enhancements/issues/12

    Pros:

>     - Familiar approach using industry standard templates.

Reviewer 3: ```suggestion
    - Familiar approach using industry standard tooling.
```

>     - Fully declarative and can be safely used in CI pipelines.

Reviewer 2: again, i'm not sure i fully understand why this is fully declarative and istioctl/operator are not.

>     - Fully declarative and can be safely used in CI pipelines.

Reviewer 3: ```suggestion
    - Native release and upgrade management 
```

>     - Native release and upgrade management.

Reviewer 3: ```suggestion
    - Helm Native release and upgrade management.
```

Reviewer: recommend commit suggestion.

    Cons:

    - Fewer checks and validations compared to `istioctl install` and Operator.
    - Some administrative tasks require more steps and have higher complexity.

>     - Complex configuration may require [kustomize](https://kubernetes.io/docs/tasks/manage-kubernetes-objects/kustomization/) or editing templates.

Reviewer 3: I'm not sure if we need to add this.

Reviewer 2: agreed, let's leave it out.
